### ISSUE TITLE
Enable type parameter for service registration methods

### ISSUE THREAD

**Issue by mattwzawislak** [grpc/grpc-dotnet#2056]
### Is your feature request related to a problem? Please describe.
Registering gRPC services is a manual process with no support for bulk registration methods. This leads to errors, missed services, and git conflicts when working on a team. 

### Describe the solution you'd like
Augment generic methods for service registration to also allow passing in a type as a parameter. Deciding what is bulk-registered should be left to the user to determine, so it is our fault if we mess something up. Having a few methods that support service registration by using `Type` would be sufficient.

For example:
````
    public void ConfigureServices(IServiceCollection services)
    {
        services.AddGrpcClient(typeof(MyProtoServiceV1.MyProtoServiceV1Client));
    }
````

This would allow bulk registration methods to exist like:

Client Service registration:
```` 
    public void ConfigureService(IServiceCollection services)
    {
        var grpcClientTypes = Assembly.GetAssembly(typeof(SomeTypeInTheProtoAssembly))
            .GetTypes()
            .Where(type => type.IsClass)
            .Where(type => type.BaseType.IsGenericType)
            .Where(type => type.BaseType.GetGenericTypeDefinition() == typeof(global::Grpc.Core.ClientBase<>));

        foreach (var grpcClientType in grpcClientTypes) 
        {
            var clientBuilder = services.AddGrpcClient(grpcClientType);

            // TODO: find interceptors linked to grpcClientType
            var interceptorTypes = Array.Empty<Type>();
            foreach (var interceptorType : interceptorTypes) 
            {
                clientBuilder.AddInterceptor(interceptorType);
            }
        }
    }
````

Server Service registration:
```` 
    public void ConfigureService(IServiceCollection services)
    {
        var grpcServerTypes = GetGrpcServerTypes();

        var serverBuilder = services.AddGrpc();
        foreach (var grpcServerType in grpcServerTypes) 
        {
            serverBuilder.AddServiceOptions(grpcServerType, options => 
            {
                // TODO: find interceptors linked to grpcServerType 
                var interceptorTypes = Array.Empty<Type>();

                foreach (var interceptorType : interceptorTypes) 
                {
                    options.Interceptors.Add(interceptorType);
                }
            });
        }
    }

    public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
    {    
        app.UseRouting();
        app.UseEndpoints(endpoints =>
        {
            var grpcServerTypes = GetGrpcServerTypes();
            foreach (var grpcServerType in grpcServerTypes) 
            {
                endpoints.MapGrpcService(grpcServerType);
            }
        });
    }

    private IEnumerable<Type> GetGrpcServerTypes()
    {
        return Assembly.GetExecutingAssembly()
            .GetTypes()
            .Where(type => type.IsClass)
            // TODO: is there a better way then BindServiceMethod? Seems sketchy
            .Where(type => type.GetCustomAttribute<global::Grpc.Core.BindServiceMethodAttribute>() is not null);         
    }
````

The methods that seem most useful to add a type parameter to are:

- `Grpc.AspNetCore.Server/GrpcEndpointRouteBuilderExtensions.MapGrpcService`
- `Grpc.AspNetCore.Server/GrpcServicesExtensions.AddServiceOptions`
  - The `AddServiceOptions` may not be do-able, but would be nice for consistency between server and client. Being able to add interceptors with bulk registration at the very least would be a win.
- `Grpc.Net.ClientFactory/GrpcClientServiceExtensions.AddGrpcClient`
- `Grpc.Net.ClientFactory/GrpcHttpClientBuilderExtensions.AddInterceptor`

### Describe alternatives you've considered
Currently my team is registering all of these options manually which is leading to mistakes and lost time. If these methods supported a `Type` as a parameter then we can register these services in bulk. 

### Additional context
I attempted a quick feasibility check on `MapGrpcService`, `AddGrpcClient` and `AddInterceptor` and felt they can all be converted to support this.

**Comment by JamesNK** [comment]
The reason it's explicit is that scanning assemblies for types isn't compatible with AOT. It also has an impact on startup performance. This is something we've observed with automatically registering all MVC controllers.

I think something that could be done to remove the need for many `MapGrpcService` calls is adding a roslyn generator. The roslyn generator would discover all the gRPC service implementations at build and create one method to register them all at once.

I don't see the need to do anything with `AddServiceOptions` because the overload that doesn't have a generic type automatically applies settings to all services.

**Comment by mattwzawislak** [comment]
> The reason it's explicit is that scanning assemblies for types isn't compatible with AOT. It also has an impact on startup performance. This is something we've observed with automatically registering all MVC controllers.

Adding in the option to register services in this manner would not impact explicit registration for those wishing to utilize AOT. Bulk registration has become a staple of many IoC frameworks, including ASP.NET MVC controllers as you mentioned. 

> I think something that could be done to remove the need for many `MapGrpcService` calls is adding a roslyn generator. The roslyn generator would discover all the gRPC service implementations at build and create one method to register them all at once.

This is a valid option that I will look into. I am initially hesitant to start on a solution like this as it would add a layer of abstraction. Not everyone on my team may be comfortable with source generators (moving off of WCF was a large enough leap as it is). I imagine other teams may be in a similar situation. The current approach is quite clear where and how registration happens and the proposed solution would still be quite direct. 

> I don't see the need to do anything with `AddServiceOptions` because the overload that doesn't have a generic type automatically applies settings to all services.

If interceptors only need to be applied to a single service type then the generic options overload would be necessary. It would be nice to discover interceptors as services are being registered.

**Comment by JamesNK** [comment]
> Adding in the option to register services in this manner would not impact explicit registration for those wishing to utilize AOT. Bulk registration has become a staple of many IoC frameworks, including ASP.NET MVC controllers as you mentioned.

ASP.NET Core MVC isn't a good pattern to follow. Assembly scanning is possible, and there is nothing stopping someone from writing a NuGet package with an extension to do it. I just don't want to include it in `Grpc.AspNetCore`.

> This is a valid option that I will look into. I am initially hesitant to start on a solution like this as it would add a layer of abstraction. Not everyone on my team may be comfortable with source generators (moving off of WCF was a large enough leap as it is). I imagine other teams may be in a similar situation. The current approach is quite clear where and how registration happens and the proposed solution would still be quite direct.

If you write something, then I'd be interested in looking at and possibly reusing the source code. A source generator to bulk register would be useful.

> If interceptors only need to be applied to a single service type then the generic options overload would be necessary. It would be nice to discover interceptors as services are being registered.

Sorry, I don't understand. If you want to discuss this further, maybe provide a scenario with some pseudo-code as an example.

**Comment by mattwzawislak** [comment]
> ASP.NET Core MVC isn't a good pattern to follow. Assembly scanning is possible, and there is nothing stopping someone from writing a NuGet package with an extension to do it. I just don't want to include it in `Grpc.AspNetCore`. 

Assembly scanning would not be a part of this project. I completely agree that this is not a good pattern for this project to provide. The sample code I provided at the top would be part of the user's code base and not your responsibility. There might be some confusion here so I'll try again. For the [`AddGrpcClient`](https://github.com/grpc/grpc-dotnet/blob/e28a43ffed112ad571788f98601674713f843507/src/Grpc.Net.ClientFactory/GrpcClientServiceExtensions.cs#L56-L61) method, currently the only way to provide a type is through the generic type parameter `TClient`:

    services.AddGrpcClient<MyGrpcClient>();
    
This suggestion was about adding a second method alongside this method that allows the type to be provided as a parameter:

    services.AddGrpcClient(typeof(MyGrpcClient));
    
 Users would be free to create their own assembly scanning and have an easy way to provide clients. For example: [`TryAddTransient<TService>()`](https://github.com/dotnet/runtime/blob/62288fd83833281ff7dfa551fe28711a2c9197c3/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/Extensions/ServiceCollectionDescriptorExtensions.cs#L163) and [`TryAddTransient(Type)`](https://github.com/dotnet/runtime/blob/62288fd83833281ff7dfa551fe28711a2c9197c3/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/Extensions/ServiceCollectionDescriptorExtensions.cs#L104-L106).
 
 Similar supporting methods could be created for [`AddInterceptor<TInterceptor>`](https://github.com/grpc/grpc-dotnet/blob/456098bcfeb9c796c19ea023b3ab3703181f1089/src/Grpc.Net.ClientFactory/GrpcHttpClientBuilderExtensions.cs#L290), [`MapGrpcService<TService>`](https://github.com/grpc/grpc-dotnet/blob/456098bcfeb9c796c19ea023b3ab3703181f1089/src/Grpc.AspNetCore.Server/GrpcEndpointRouteBuilderExtensions.cs#L38) and [`AddServiceOptions<TService>`](https://github.com/grpc/grpc-dotnet/blob/fb315ca62b5c1cd805906088fcd4354b00c3fed1/src/Grpc.AspNetCore.Server/GrpcServiceExtensions.cs#L42).

Hopefully I helped clarify the expectations here. The idea was not to introduce assembly scanning to this project, but to provide an alternative way to invoke 4 existing methods. 

> If you write something, then I'd be interested in looking at and possibly reusing the source code. A source generator to bulk register would be useful.

I can try but it may take a while. I am quite busy at work at the moment and unfortunately don't have much free time. I'll try to put some personal time aside to try and create a working source generator.

**Comment by JamesNK** [comment]
It's not just a matter of adding a `Type` overload. Generic types flow through the entire system.

The only way to realistically do this is have `AddGrpcClient(Type)` use reflection to invoke the generic method: `addGrpcClientMethodInfo.MakeGenericMethod(type).Invoke(null)`. I want to avoid adding methods like that because of the unexpected perf overhead of reflection.

**Comment by mattwzawislak** [comment]
I took a look at the code before suggesting this and felt it was possible to do this without reflection and without duplicating the entire inner-workings. There may be some trickiness in handling the `RequiresDynamicCode` but that is about it. Those warnings could just bubble up to the public-facing API. I was able to fork the project and get `AddGrpcClient(Type)` working. No guarantees the rest would work though. 

I get the feeling this is not going to happen, so I will close this out.
